File: solutions/Combination-Sum/Combination-Sum.py


class Solution:
    def dfs(self,candidates,target,pos,vec,ans):
        if target==0:
            vec2=vec[:]
            ans.append(vec2)
            return 
        for i in range(pos,len(candidates)):
            if target<candidates[i]:
                return 
            vec.append(candidates[i])
            self.dfs(candidates,target-candidates[i],i,vec,ans)
            vec.pop()
    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def combinationSum(self, candidates, target):
        candidates.sort()
        vec=[]
        ans=[]
        self.dfs(candidates,target,0,vec,ans)
        return ans
 

File: solutions/Combination-Sum/Combination-Sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
  candidates.sort((x, y) => x < y ? -1 : 1);
  const ans = [];
  const dfs = (vec, tar, pos) => {
    if (tar === 0) {
      ans.push(vec.slice());
      return;
    }
    for (let i = pos; i < candidates.length; i++) {
      if (tar < candidates[i]) {
        return;
      }
      vec.push(candidates[i]);
      dfs(vec, tar - candidates[i], i);
      vec.pop();
    }
  };
  dfs([], target, 0);
  return ans;
};